strawman delegate implementation

acoburn opened the pull request:
This doesn't even pretend to be a correct implementation (it doesn't even consider type-based accessToClass resources), but it is a start. Please feel free to tear it apart!

awoods commented on src/main/java/org/fcrepo/auth/webac/WebACAuthorizationDelegate.java:
```diff
@@ -37,10 +49,45 @@
+
+        final List<URI> actionURIs = actionsAsURIs(actions);
+
+        final Optional<URI> effectiveACL = getEffectiveAcl(new FedoraResourceImpl(userSession.getNode(absPath)));
```
We may need an "internal-session" as the `userSession` will likely not have permissions to read the ACL resources.

awoods commented on src/main/java/org/fcrepo/auth/webac/WebACAuthorizationDelegate.java:
```diff
@@ -37,10 +49,45 @@
-        final boolean permit = false;
+
+        // This is not correct -- we should get it from the container or header, etc
+        final String agent = userSession.getUserID();
```
This method could be made `protected`:
https://github.com/fcrepo4/fcrepo-module-auth-rbacl/blob/master/fcrepo-auth-roles-common/src/main/java/org/fcrepo/auth/roles/common/AbstractRolesAuthorizationDelegate.java#L141

awoods replied:
As a note: "group" principals are available as jcr.Session attributes set here:
https://github.com/fcrepo4/fcrepo4/blob/master/fcrepo-auth-common/src/main/java/org/fcrepo/auth/common/ServletContainerAuthenticationProvider.java#L174

awoods commented:
As a side note, we will need a WebAC implementation of `AccessRolesProvider`:
https://github.com/fcrepo4/fcrepo-module-auth-rbacl/blob/master/fcrepo-auth-roles-common/src/main/java/org/fcrepo/auth/roles/common/AbstractRolesAuthorizationDelegate.java#L53

acoburn commented:
I would also note that I used `Optional<boolean>` instead of a more straight-forward `Predicate` because we actually need to differentiate between `null`, `true` and `false`, where `true` means "permit the operation", `false` means "deny the operation" and `null` means try the next authorization mechanism (starting with `user`, then moving to `group` and finally trying `rdf:type` for `acl:accessToClass` properties). If none of those provide a non-null response, then deny the request (i.e. `permit = false`).

whikloj commented on src/main/java/org/fcrepo/auth/webac/WebACAuthorizationDelegate.java:
```diff
@@ -15,6 +15,17 @@
 package org.fcrepo.auth.webac;
 
+// The WEBAC_HAS_ACL variable does not exist (we don't even have a namespace for it yet).
+import static org.fcrepo.auth.webac.URIConstants.WEBAC_HAS_ACL;
```
Could we use acl:accessControl for this? From the ontology
```
<rdf:Property rdf:about="http://www.w3.org/ns/auth/acl#accessControl">
    <comment>The Access Control file for this information resource.
    This may of course be a virtual resorce implemented by the access control system.
    Note also HTTP's header  Link:  foo.meta ;rel=meta can be used for this.</comment>
    <domain rdf:resource="http://www.w3.org/2006/gen/ont#InformationResource"/>
    <label>access control</label>
    <range rdf:resource="http://www.w3.org/2006/gen/ont#InformationResource"/>
    <subPropertyOf rdf:resource="http://www.w3.org/2000/01/rdf-schema#seeAlso"/>
</rdf:Property>
```

awoods replied:
@whikloj, that property seems appropriate.

acoburn replied:
:+1: nice call

awoods commented:
@acoburn, are we actually planning on hijacking `acl:accessToClass`? Do you not have concerns about then changing gears in phase two towards dereferencing those URIs to get the membership of those "groups"?
Unless we do not have such phase-two plans.

acoburn commented:
@awoods If you look at the examples in the [WebAC wiki](http://www.w3.org/wiki/WebAccessControl), you will see that `acl:accessToClass` is used, e.g. to reference `foaf:Document`. The range is any `rdfs:Class` of resources, so using that to reference `ex:AdminImage` or `ex:ArchivalObject` as in:

    <> acl:accessToClass ex:ArchivalObject, ex:AdminImage ;
        acl:agent <http://myinstitution.edu/groups/admins> ;
        acl:mode acl:Read, acl:Write .

Do you feel that this diverges from what the WebACL document lays out?

awoods commented:
@acoburn, nevermind... I was mentally hooked on `acl:agentClass` not `acl:accessToClass`. Disregard.

acoburn commented on src/main/java/org/fcrepo/auth/webac/WebACAuthorizationDelegate.java:
```diff
@@ -37,10 +49,45 @@
+        map.put("DELETE", WEBAC_MODE_WRITE);
+        map.put("PATCH", WEBAC_MODE_WRITE);
+        map.put("OPTIONS", WEBAC_MODE_READ);
+        actionMap = Collections.unmodifiableMap(map);
```
I have no idea if this matches any type of reality. Is the `actions` array (below) composed of Strings like "GET", or "POST"? Without knowing any better, that's what I'm assuming.

ajs6f replied:
Incidentally, if @acoburn is correct, this is one reason why we don't want to have resources created or updated or deleted by requests to URIs that aren't their own.

awoods commented:
Are we scrapping this PR in favor of a new one? or is there a plan for closure today?

acoburn commented:
I'm still working on this PR, though it is dependent on @whikloj and the creation of a WebACRolesProvider interface

whikloj commented:
Sorry, but I thought @mohideen was going to work on the WebACRolesProvider implementation as he was already working on the AuthHandler interface, I was working on the WebACAuthenticationImpl. Did I get that mixed up?

acoburn commented:
Added unit tests with the provided TTL resources. Added more debugging log output. Need to add more comments, and have support for acls residing on ancestor nodes.

acoburn commented:
This PR is now deprecated in favor of #28 (and other yet-to-be-issued PRs)

awoods commented:
@acoburn, please feel free to close this PR if that is the intent of your previous comment.
